mv: handle mismatched base/view replica count caused by RF change

wmitros opened the pull request:
During an ALTER KEYSPACE statement execution where a table with a view
is present, we need to perform tablet migrations for both tables.
These migrations are not synchronized, so at some point the base may
have a different number of non-pending replicas than the view. Because
of that, we can't pair them correctly. If there is more non-pending
base replicas than view replicas, we don't need to do anything because
the view replica that didn't finish migrating is a pending replica
and will get view updates from all base replicas. But if there is more
non-pending view replicas than base replicas, we may currently lose
view updates to the new view replica.

This patch adds a workaround for this scenario. If after one migration
we have too more non-pending view replicas than base replicas, we add
it to the pending replica list so that it gets an update anyway.

This patch will also take effect if the base and view replica counts
differ due to some other bug. To track that, a new metric is added
to count such occurrences.

This patch also includes a test for this exact scenario, which is enforced by an injection.

Fixes https://github.com/scylladb/scylladb/issues/21492

wmitros commented:
The issue is specific to views with tablets which are experimental now, so not backporting

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17532/) | :white_check_mark: | i-017831bab5c6de46b |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17532/) | :x: | spider4.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17532/) | :x: | spider5.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17532/) | :x: | spider7.cloudius-systems.com |

#### Build Details:
- Duration: 26 min

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17533/) | :white_check_mark: | sif |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17533/) | :x: | spider8.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17533/) | :x: | spider1.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17533/) | :x: | spider6.cloudius-systems.com |

#### Build Details:
- Duration: 11 min

avikivity commented on db/view/view.cc:
```diff
@@ -1781,8 +1781,23 @@ bool should_generate_view_updates_on_this_shard(const schema_ptr& base, const lo
+// We don't need to return the second replica if we have more base replicas than
+// view replicas - in that case the corresponding view replica will get the
+// update anyway, because at that point it must be pending.
+std::pair<std::optional<locator::host_id>, std::optional<locator::host_id>>
```
Better to create a struct with names.

wmitros commented:
In the rebase:
* fixed the wrong paired base replica
* removed a redundant check of a left unpaired replica
* used a struct instead of a pair for the endpoints returned by `get_view_natural_endpoint`

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17582/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17582/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17582/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/network_topology_strategy_test
:small_blue_diamond: cluster/mv/test_mv_topology_change
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5436/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8604/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16687/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17582/)
#### Build Details:
- Duration: 7 hr 55 min
- Builder: spider8.cloudius-systems.com

bhalevy commented on db/view/view.cc:
```diff
@@ -1962,6 +1993,24 @@ get_view_natural_endpoint(
+        if (base_endpoints.size() < view_endpoints.size() - 1) {
+            // We only expect one extra replica to appear due to an RF change. If there's more, that's an error,
+            // but we'll still perform updates to the paired and last replicas to minimize degradation.
+            vlogger.error("There's too many view endpoints for base-view pairing. View updates may get lost on view_endpoints={}",
```
nit: this should probably be a warning, not an error, and also should be "There are" rather "There's" since view endpoints are plural.

bhalevy replied:
Also what can/should the user do about it?
What can we do about it if someone opens a support ticket seeing this scary message?

wmitros replied:
If support gets a ticket with this and the user relies on high base-view consistency (which we don't really provide), they should perform repair on the view.
This message shouldn't ever appear unless we have a bug, so if it appears anywhere we should investigate and fix it.
If it appears, it also doesn't necessarily mean there's an inconsistency, but that we detected a possibility of one (which we're also not that good at), so if there's only one or few such messages, if the user is okay with a few inconsistencies in the view, they may not need to do anything

bhalevy commented on db/view/view.cc:
```diff
@@ -1962,6 +1993,24 @@ get_view_natural_endpoint(
+        // to send an additional update to it.
+        ++cf_stats.total_view_updates_due_to_replica_count_mismatch;
+        no_pairing_replica = view_endpoints.back();
+        if (base_endpoints.size() < view_endpoints.size() - 1) {
```
@Deexie will that work when changing RF from 0 -> N incrementally?
Do we do the increments of 1 replica at a time on all tables in the keyspace including base and views?

Deexie commented on db/view/view.cc:
```diff
@@ -2064,11 +2121,19 @@ future<> view_update_generator::mutate_MV(
+        if (no_pairing_endpoint) {
+            // The no_pairing_endpoint can appear during a replication factor increase when the base tablet migration finished after
+            // the corresponding (for current token) view tablet migration. In this case, the view replica is no longer
+            // in migration (so it's not present in the pending endpoints list), but the base replica still is. We don't pair pending
```
After RF 0->N changes the base replica won't necessarily be in migration - it will start the migration eventually.

wmitros commented on db/view/view.cc:
```diff
@@ -1962,6 +1995,24 @@ get_view_natural_endpoint(
+        // Because of this, a view replica might not get paired with any base replica, so we need
+        // to send an additional update to it.
+        ++cf_stats.total_view_updates_due_to_replica_count_mismatch;
+        no_pairing_replica = view_endpoints.back();
```
Actually, it's chosen only from the current DC which means that if the view finishes streaming before the base tablet in a 0->1 RF change in a DC, no base will currently send updates to it. At the same time, I'd like to avoid cross-DC view updates...
@Deexie regarding the comment
> After RF 0->N changes the base replica won't necessarily be in migration - it will start the migration eventually.

Does it mean that after the RF 0->N changes we won't start rebuilding tablets of all tables at the same time? If so, would it be feasible to rebuild tablets of base tables before rebuilding view tablets? Or would it be too restricting/complicated?

Deexie replied:
> Actually, it's chosen only from the current DC which means that if the view finishes streaming before the base tablet in a 0->1 RF change in a DC, no base will currently send updates to it. At the same time, I'd like to avoid cross-DC view updates... @Deexie regarding the comment
> 
> > After RF 0->N changes the base replica won't necessarily be in migration - it will start the migration eventually.
> 
> Does it mean that after the RF 0->N changes we won't start rebuilding tablets of all tables at the same time?

Yes, exactly.

> If so, would it be feasible to rebuild tablets of base tables before rebuilding view tablets? Or would it be too restricting/complicated?

We can easily check if a tablet is of a base or view table, right? If so, then yes, it will make it more complicated, but it should be possible.

wmitros replied:
> We can easily check if a tablet is of a base or view table, right? If so, then yes, it will make it more complicated, but it should be possible.

We only need to look at the schema or `table` so should be relatively simple. If we add that restriction there, this entire patch won't be necessary. Otherwise I suppose we'll have to rely on cross-DC updates for the case where there the first base replica in the DC didn't finish streaming yet but the view replica did

Deexie commented:
> @Deexie What about multiple DCs? Will DCs be allowed to increment their RF independently, or will it be serialized?

> If yes, then perhaps it would be better to return a utils::small_vector instead of a single unpaired replica.

It will be serialized

piodul commented:
> #### Failed Tests (6/75808):
> 
>     * [test_mv_first_replica_in_dc[base].dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19122/testReport/junit/cluster.mv/test_mv_topology_change/dev___test_py___dev___test_mv_first_replica_in_dc_base__dev_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_first_replica_in_dc%5Bbase%5D.dev.1)
> 
>     * [test_mv_first_replica_in_dc[base].dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19122/testReport/junit/cluster.mv/test_mv_topology_change/dev___test_py___dev___test_mv_first_replica_in_dc_base__dev_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_first_replica_in_dc%5Bbase%5D.dev.2)
> 
>     * [test_mv_first_replica_in_dc[base].dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19122/testReport/junit/cluster.mv/test_mv_topology_change/dev___test_py___dev___test_mv_first_replica_in_dc_base__dev_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_first_replica_in_dc%5Bbase%5D.dev.3)
> 
>     * [test_mv_first_replica_in_dc[base].debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19122/testReport/junit/cluster.mv/test_mv_topology_change/debug___test_py___debug___test_mv_first_replica_in_dc_base__debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_first_replica_in_dc%5Bbase%5D.debug.1)
> 
>     * [test_mv_first_replica_in_dc[base].debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19122/testReport/junit/cluster.mv/test_mv_topology_change/debug___test_py___debug___test_mv_first_replica_in_dc_base__debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_first_replica_in_dc%5Bbase%5D.debug.2)
> 
>     * [test_mv_first_replica_in_dc[base].debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/19122/testReport/junit/cluster.mv/test_mv_topology_change/debug___test_py___debug___test_mv_first_replica_in_dc_base__debug_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_mv_first_replica_in_dc%5Bbase%5D.debug.3)
> 

@wmitros Looks like a legitimate issue with the PR:

```
15:17:27          # Confirm that the problematic scenario occurred
15:17:27          replica_mismatches = await get_replica_mismatches()
15:17:27  >       assert replica_mismatches > replica_mismatches_before
15:17:27  E       assert 0 > 0
```

wmitros commented:
> #### 🔴 CI State: FAILURE

The failure was due to an intranode tablet migration which we confused with the migration between dc's. I fixed it using '--smp 1'

wmitros commented:
The issue should have been fixed by setting '--smp 1' but I committed the `--smp 2` used for testing. Should be fine now

wmitros commented on test/cluster/mv/test_mv_topology_change.py:
```diff
@@ -201,3 +203,131 @@ async def test_mv_write_to_dead_node(manager: ManagerClient):
+@skip_mode('release', 'error injections are not supported in release mode')
+async def test_mv_first_replica_in_dc(manager: ManagerClient, delayed_replica: str):
+    servers = []
+    servers.append(await manager.server_add(cmdline=['--smp', '1'], config={'rf_rack_valid_keyspaces': False}, property_file={'dc': f'dc1', 'rack': 'myrack1'}))
+    servers.append(await manager.server_add(cmdline=['--smp', '1'], config={'rf_rack_valid_keyspaces': False}, property_file={'dc': f'dc2', 'rack': 'myrack1'}))
```
I added a comment explaining why it's needed in the "new_dc" case.
We don't need it in the `test_mv_rf_change` case because we use 2 nodes per rack there and if the base and view tablets get allocated on the same node and shard initially, we'll wait until we finish the migration of one of these tablets to the other node in the rack in `replicas_balanced()`

wmitros commented on test/cluster/mv/test_mv_topology_change.py:
```diff
@@ -201,3 +203,131 @@ async def test_mv_write_to_dead_node(manager: ManagerClient):
+    await asyncio.gather(*[manager.api.message_injection(s.ip_addr, f"block_tablet_streaming") for s in servers])
+    await ks_fut
+
+    # Confirm that the problematic scenario occurred
+    replica_mismatches = await get_replica_mismatches()
+    assert replica_mismatches > replica_mismatches_before
```
In both test cases we perform 1 migration (rebuild) of base and view tablet each. A failure should be counted only for the migration that finishes later, so we should have 1 failure across all replicas.
So I updated the test to check that the mismatches increased by 1

lsfreitas commented:
CI failed on this PR due to an issue introduced and already reverted causing a wrong status. Please re-trigger CI

lsfreitas commented:
CI failed on this PR due to an issue introduced and already reverted causing a wrong status. Please re-trigger CI

piodul commented:
```
        # Confirm that the problematic scenario occurred
        replica_mismatches = await get_replica_mismatches()
>       assert replica_mismatches == replica_mismatches_before + 1
E       assert 2.0 == (0 + 1)
```

@wmitros Please take a look at the failure first before retriggering, it looks legitimate.

wmitros commented:
> ```
>         # Confirm that the problematic scenario occurred
>         replica_mismatches = await get_replica_mismatches()
> >       assert replica_mismatches == replica_mismatches_before + 1
> E       assert 2.0 == (0 + 1)
> ```
> 
> @wmitros Please take a look at the failure first before retriggering, it looks legitimate.

Looks like with the new view build coordinator, the view building may not finish before we test the insert. So I added the `wait_for_view` in the rebase.
I pushed the change, but the `wait_for_view` actually times out on my machine. Looks like if we specify RF for multiple DCs, the view building coordinator doesn't mark the view building as finished in `view_build_status_v2`. CC @Jadw1

Jadw1 commented:
> > ```
> >         # Confirm that the problematic scenario occurred
> >         replica_mismatches = await get_replica_mismatches()
> > >       assert replica_mismatches == replica_mismatches_before + 1
> > E       assert 2.0 == (0 + 1)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @wmitros Please take a look at the failure first before retriggering, it looks legitimate.
> 
> Looks like with the new view build coordinator, the view building may not finish before we test the insert. So I added the `wait_for_view` in the rebase. I pushed the change, but the `wait_for_view` actually times out on my machine. Looks like if we specify RF for multiple DCs, the view building coordinator doesn't mark the view building as finished in `view_build_status_v2`. CC @Jadw1

It looks like problem in view building coordinator. One task was aborted due to tablet migration but it was never updated after the migration was finished (and it happens consistently). 
```
DEBUG 2025-09-03 21:17:03,760 [shard 0: gms] view_building_coordinator - Generating abort mutations for tasks for table 92015dd0-88fa-11f0-b827-28c5d88916e9 on replica 516b7ee1-1ca5-4c93-90f8-7ee272cce92c:0 and last token 9223372036854775807
DEBUG 2025-09-03 21:17:03,760 [shard 0: gms] view_building_coordinator - Aborting task 92fa15b3-88fa-11f0-9235-62e4d4e6d76f
```
I'll look deeper into this tomorrow.

Jadw1 commented:
@wmitros I posted https://github.com/scylladb/scylladb/pull/25916, with this fix the view building finishes in multi DC setup as in your reproducer.
